Final Project Submission: Sterile Processing Tracker (SPT)

bbeare22 opened the pull request:
This PR contains my final project for the TripleTen Software Engineering program.

✨ Features
- Machines: add/edit/delete washers, sterilizers, ultrasonics
- Maintenance: descale (washers/ultrasonics), daily inspection + quarterly cleaning (sterilizers)
- Sterilizer cycles: track loads, items, result, staff, spore test details
- Dashboard: KPIs, today’s cycles, recent maintenance
- Machine detail: per-machine maintenance + cycles
- CSV export for maintenance + cycles
- Auth: register/login/logout with JWT
- UI polish: toasts, skeletons, consistent card layouts

📦 Includes:
- Seed script with demo user (`demo@spt.app` / `demo123`)
- .env.example files for server + client
- Production build instructions
- Screenshots

bbeare22 committed:
feat: scaffold Vite React app

bbeare22 committed:
feat(ui): AppShell layout, tokens, router, base pages

bbeare22 committed:
feat(ui): Card and KPI components + dashboard layout

bbeare22 committed:
feat(recalls): express proxy + live OpenFDA recalls table

bbeare22 committed:
Updated title

bbeare22 committed:
Fixed gitignore

bbeare22 committed:
chore: add root .gitignore and clean node_modules

bbeare22 committed:
Update to project

bbeare22 committed:
feat(maintenance): form shell with mL↔oz converter and validation

bbeare22 committed:
feat(machines): list with filters, MachineCard, and detail route

bbeare22 committed:
feat(ui): color-code overdue list and add days-since chip on MachineCard

bbeare22 committed:
feat(ui): add glowing status dot to MachineCard

bbeare22 committed:
feat(ui): reuse status dots across Dashboard and MachineCard

bbeare22 committed:
feat(recalls): add brand filter, loading skeleton, and empty state

bbeare22 committed:
feat(machines): read from /api/machines and support _id in UI

bbeare22 committed:
feat(maintenance): backend routes + connect form to API and auto-update lastDescaleAt

bbeare22 committed:
feat(dashboard): fetch live machines and add Log maintenance shortcut for overdue items

bbeare22 committed:
feat(dashboard): add Recent Maintenance card alongside Overdue Descales and keep Recent Cycles

bbeare22 committed:
Add employeeId and sterilizationNumber to user model, register page, and sidebar display

bbeare22 committed:
Small styling update on Login/Register/Logout

bbeare22 committed:
feat(ui): toast notifications, loading skeletons, and unified card headers

bbeare22 committed:
feat(history): add CSV export with toast notifications to maintenance history

bbeare22 committed:
Added 404 not found page and also updated the about page

bbeare22 committed:
Polish UI: fix MachineForm & Maintenance form overflow, button styles

bbeare22 committed:
Added cycle integration and a few updates

bbeare22 committed:
Add sterilizer cycles integration, cycle history page, and UI polish

bbeare22 committed:
feat: add sterilizer cycles logging, sterilizer-specific maintenance (daily inspection & quarterly cleaning), UI polish for forms and CSV export

bbeare22 committed:
chore: add env examples

bbeare22 committed:
updated readme

bbeare22 committed:
Small fix/ polish

bbeare22 committed:
Add screenshots and update README

bbeare22 committed:
screenshot fix

bbeare22 committed:
small fix

bbeare22 committed:
Add screenshots and fix README paths

bbeare22 committed:
final check before submit

gennady-bars commented on README.md:
```diff
@@ -1 +1,216 @@
-# Sterile Processing Tracker
+# Sterile Processing Tracker (SPT)
+
```
* You need to **deploy** the app, and show the link in `Readme`

gennady-bars commented on apps/client/src/pages/CyclesHistory.jsx:
```diff
@@ -0,0 +1,181 @@
+  );
+}
+
+/* ---- styles (matches MaintenanceHistory look) ---- */
+const headerBar = {
+  display: "flex",
+  justifyContent: "space-between",
+  alignItems: "center",
+  marginBottom: 16,
+};
+
+const linkBtn = {
+  color: "#cbd5e1",
+  textDecoration: "none",
+  border: "1px solid var(--color-border)",
+  borderRadius: 12,
+  padding: "8px 12px",
+};
+
+const btnGhost = {
```
According to the criteria, you need to use `BEM` classes. You should store them in `.css` files

gennady-bars commented on apps/client/src/context/AuthContext.jsx:
```diff
@@ -0,0 +1,73 @@
+    employeeId,
+    sterilizationNumber,
+  }) {
+    const r = await apiFetch("/api/auth/register", {
+      method: "POST",
+      body: JSON.stringify({
+        email,
+        password,
+        name,
+        employeeId,
+        sterilizationNumber,
```
All api functions should be stored in `api.js` and (or) `auth.js`, to separate the api logic from the components

gennady-bars replied:
Please, move all api functions into `api.js` or `auth.js`, to separate the logic

gennady-bars commented on apps/client/src/context/AuthContext.jsx:
```diff
@@ -0,0 +1,73 @@
+        sterilizationNumber,
+      }),
+    });
+    if (!r.ok) throw new Error((await r.json()).error || `HTTP ${r.status}`);
```
```js
    if (!r.ok) throw new Error((await r.json()).error || `HTTP ${r.status}`);


```

The code of `ok` checking is repeated in **every** request of the project. To avoid it, you need to make a special `checkResponse`  function that will check any response from the server if it’s ok or not.


```js
function checkResponse(res) {
        // here is the code of the checking
    }

```

And now you can replace redundant repeating with the only line:


```js
 .then(checkResponse)

```

Please, pay attention that you need to pass only the **reference** for the function rather than the call. 

----------------------------------------------------------------

**COULD BE IMPROVED**

You can make a special function for fetching and checking responses **not to duplicate** it in every request:

```js

  function request(url, options) {
    return fetch(url, options).then(checkResponse)
  }

```

And now you can safely replace `fetch` with `request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars replied:
Please, address this remark as well

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Brett!

Great job on submitting your project! This time you’ve implemented `Api` integration. So you can create your own simple sites from now on! 

Please, scroll down the page on **github** to see my remarks right in the code

**Parts I really liked in your project:**

* Well-structured valid markup
* The hierarchy of headings is correct
* No horizontal scrolling on desktop resolutions
* Great semantic tags
* All `img` tags have descriptive `alt` attributes
* Links have `hover` effect


There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code and the comments below:
* You need to **deploy** the app, and show the link in `Readme`. The deployed app should work
* According to the criteria, you need to use `BEM` classes. You should store them in `.css` files
* Please, create `ModalWithForm` for any modals with a form, to avoid markup duplication. At least the `form` tag, `close icon` should be there if you have different designs for any form
* All api functions should be stored in `api.js` and (or) `auth.js`, to separate the api logic from the components
* The code of `ok` checking is repeated in **every** request of the project. To avoid it, you need to make a special `checkResponse`  function that will check any response from the server if it’s ok or not.


There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

bbeare22 committed:
Refactor date handling: shared helpers + aligned forms

bbeare22 committed:
Refactor: use apiFetch everywhere; date helpers; small cleanups

bbeare22 committed:
Refactor: centralize API via apiFetch + auth utils; date helpers

bbeare22 committed:
Add ErrorBoundary and wrap app to avoid hard crashes

bbeare22 committed:
UI: add accessible Modal and use it for Add/Edit Machine

bbeare22 committed:
refactor: move inline styles to separate CSS files across components and pages

bbeare22 committed:
Preparing for deploy

bbeare22 committed:
Updated ReadME

bbeare22 commented:
Stage 1 Resubmission: Sterile Processing Tracker (SPT)

This resubmission addresses reviewer feedback:
- Centralized API calls in utils
- Added ErrorBoundary
- Moved inline styles into CSS modules
- Deployed backend & frontend to Render
- Updated README with live links and walkthrough

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!

bbeare22 committed:
feat: add RBAC for maintenance/cycles + supervisor-only machine management

bbeare22 committed:
feat: update spores, cycles, maintenance logging
- Dashboard: added Pending Spore KPI and reordered cards
- SporeQueue: added load #, verified by/at columns
- LogCycle: simplified spore fields (removed verify upfront)
- Maintenance: track initials (createdBy), updated schema & routes
- MachineDetail: show Logged By for maintenance and cycles
- MaintenanceHistory: show Performed By initials
- CyclesHistory: show Load Staff, Unload Staff, Verified By initials

bbeare22 committed:
Added PM page

bbeare22 committed:
Add month-based CSV exports, fix dashboard KPIs with start/end filters, align recent cycles display

bbeare22 committed:
Added Decon page

bbeare22 committed:
Updated decon Page

bbeare22 committed:
Added control BI testing

bbeare22 committed:
Updated maintenance, added washer daily rack test/ washer weekly maintenance

bbeare22 committed:
Added transport page for logging transport trips and fuel use

bbeare22 committed:
Updated transport page to show monthly trip stats

bbeare22 committed:
Added audit section

bbeare22 committed:
Added server side audit function

bbeare22 committed:
Add PDF exports and daily compliance reminders job

bbeare22 committed:
Clean reset

bbeare22 committed:
Add tablet friendly pass

bbeare22 committed:
Small fixes and polishing

bbeare22 committed:
Small index.js fix

bbeare22 committed:
Add branded logo, favicon, and manifest for SPT + report enhancements

bbeare22 committed:
Fix: correct maintenance initials mapping in reports (PDF & CSV)

bbeare22 committed:
Stop tracking env files; keep only sanitized .env.example

bbeare22 committed:
Stop tracking env files; keep only sanitized .env.example

bbeare22 committed:
add sanitized server .env.example

bbeare22 committed:
wrap routes with ErrorBoundary

bbeare22 committed:
small push

bbeare22 committed:
client: remove unused boilerplate and dead files

bbeare22 committed:
chore: add lightweight logger; quiet common console noise

bbeare22 committed:
chore: add root scripts and Prettier config; enable format-on-save

bbeare22 committed:
chore: add ESLint (flat), lint scripts, and optional pre-commit hooks

bbeare22 committed:
Update README.md

README.md update

bbeare22 committed:
Update for demo mode

bbeare22 committed:
Add MIT License
